sources/filesystem/gitpythonfs/gitpythonfs/core.py

```diff
@@ -0,0 +1,210 @@
+                "gitpythonfs://[path-to-repo:][ref@]path/to/file" so that repo
+                and/or ref can be passed in the URL instead of arguments. (The
+                actual file path should not contain "@" or ":"). Examples:
+                    When instantiating GitPythonFileSystem:
+                    /some_folder/my_repo
+                    When calling open(), open_files() etc:
+                    gitpythonfs:///some_folder/my_repo:path/to/intro.md
+                    gitpythonfs:///some_folder/my_repo:mybranch@path/to/intro.md
+            ref (str): (To be implemented). A branch, tag or commit hash to use.
+                Defaults to head of the local repo.
+        """
+        super().__init__(**kwargs)
+        self.repo_path = path
+        self.repo = git.Repo(self.repo_path)
+
+    @classmethod
+    def _strip_protocol(cls, path: str) -> str:
+        path = super()._strip_protocol(path).lstrip("/")
+        if ":" in path:
+            path = path.split(":", 1)[1]
+        if "@" in path:
+            path = path.split("@", 1)[1]
+        return path.lstrip("/")
+
+    # ToDo support arguments in url, like this example from git fsspec implementation:
+    @staticmethod
+    def _get_kwargs_from_urls(path: str) -> Dict[str, str]:
+        if path.startswith("gitpythonfs://"):
+            path = path[14:]
+        out = {}
+        if ":" in path:
+            out["path"], path = path.split(":", 1)
+        if "@" in path:
+            out["ref"], path = path.split("@", 1)
+        return out
+
+    def _git_type_to_file_type(self, object: git.Object) -> str:
+        if isinstance(object, git.Blob):
+            return "file"
+        elif isinstance(object, git.Tree):
+            return "directory"
+        else:
+            msg = f"There is no fileystem object type corresponding to Git object type: {type(object).__name__}"
+            raise TypeError(msg)
+
+    def _details(
+        self, object: git.Object, include_committed_date: bool = True
+    ) -> Dict[str, Union[str, int]]:
+        """
+        Retrieves the details of a Git object.
+
+        Args:
+            object (git.Object): The Git object to retrieve details for.
+            include_committed_date (bool, optional): Whether to include the committed date. Defaults to True.
+                Getting the committed date is an expensive operation and will slow down
+                walk(), a method that is extensively used by fsspec for find(), glob() etc.
+
+        Returns:
+            dict: A dictionary containing the details typical for fsspec.
+        """
+        # commit=next(self.repo.iter_commits(paths=object.path, max_count=1))
+        details = {
+            "name": object.path,
+            "type": self._git_type_to_file_type(object),
+            "mime_type": object.mime_type if isinstance(object, git.Blob) else None,
+            "size": object.size,
+            "hexsha": object.hexsha,
```

Comment:
1. we can call it `hex`
2. we can also keep `mode` to make it compatible with git://